### PR TITLE
[core][iOS] Provide Exception.code to JavaScript (as Error.code) for all types of functions

### PR DESCRIPTION
**Pull request by cltnschlosser**
# Why

https://github.com/expo/expo/issues/27959

# How

- FunctionCallException uses code from its cause if there is one. (This is sufficient to pass the code for all AsyncFunctions)
- Synchronous functions now create a custom Error payload which includes `code`.

# Test Plan

Early commit used tests using `apps/test-suite`. This commit includes unit tests for synchronous functions.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by tsapeta — CHANGES_REQUESTED** [review]
The changes for error codes looks good to me, however I'm a bit concerned about adding another standalone package just for these tests. I think I'll need to set up a better environment to add native modules to test-suite just for testing. Probably best if they are local modules.

That being said, can you remove the `expo-modules-exception-test` package from this PR and keep just the changes that are needed for error codes?

If you do want to provide some tests, you could add a simple test for `FunctionCallException` here: https://github.com/expo/expo/blob/main/packages/expo-modules-core/ios/Tests/ExceptionsSpec.swift (without going through the native module). Might be a good idea to add it inside `describe("FunctionCallException") {}`.

For a test that goes through different layers (DSL, native module, runtime, JS), take a look at these tests: https://github.com/expo/expo/blob/31f646a5fe93f36d62b262fab5f3edd1e5287feb/packages/expo-modules-core/ios/Tests/FunctionSpec.swift#L245
There is a mocked native module already. Sadly, we don't have an easy way to test AsyncFunction from unit tests yet – we'll need to write some helpers to conveniently work with promises, so that will be enough if you add a test only for a sync function.

These tests can be run from Xcode in bare-expo project. See below:
Screenshot 2024-04-12 at 09 47 43

Let me know if you need any more assistance on this!

**Comment by cltnschlosser** [comment]
Okay, @tsapeta this should be ready for review, let me know what you think.

File: packages/expo-modules-core/CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@
 - JS object of the native module is now an instance of the `NativeModule` class that inherits from `EventEmitter`. ([#27510](https://github.com/expo/expo/pull/27510) by [@tsapeta](https://github.com/tsapeta))
 - [iOS] Exposed a function on the runtime to schedule some work with synchronized access to JS. ([#27567](https://github.com/expo/expo/pull/27567) by [@tsapeta](https://github.com/tsapeta))
 - [iOS] `OnStartObserving` and `OnStopObserving` can now be attached to a specific event. ([#27766](https://github.com/expo/expo/pull/27766) by [@tsapeta](https://github.com/tsapeta))
+- [iOS] Provide Exception.code to javascript (as Error.code) for all types of functions. ([#27960](https://github.com/expo/expo/pull/27960) by [@cltnschlosser](https://github.com/cltnschlosser))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:27]:
```suggestion
- [iOS] Provide Exception.code to JavaScript (as Error.code) for all types of functions. ([#27960](https://github.com/expo/expo/pull/27960) by [@cltnschlosser](https://github.com/cltnschlosser))
```

File: packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm [hunk 2]
```diff
@@ -112,7 +112,20 @@ - (nonnull EXJavaScriptObject *)createSyncFunction:(nonnull NSString *)name
       if (error == nil) {
         return expo::convertObjCObjectToJSIValue(runtime, result);
       } else {
-        throw jsi::JSError(runtime, [error.userInfo[@"message"] UTF8String]);
+        // `expo::makeCodedError` doesn't work during unit tests, so we construct Error and add a code,
+        // instead of using the CodedError subclass.
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm:116]: I'll take a look what was wrong 👍 For now that's fine.

**Review by tsapeta — APPROVED** [review]
Looks good, thanks for taking care of that! Just apply my suggestion in the changelog and it's ready to go 👍

**Comment by cltnschlosser** [comment]
> Looks good, thanks for taking care of that! Just apply my suggestion in the changelog and it's ready to go 👍

Not that I expect you to remember my preferences against everyone else’s, but I don’t mind if you accept the change yourself :)

**Comment by tsapeta** [comment]
Would you mind if it comes out in SDK 51 (in less than a month, beta release pretty soon)? In other words, do you need this to be backported to SDK 50?

**Comment by cltnschlosser** [comment]
SDK 51 is okay

**Pull request merged**